CMakeLists.txt

```diff
@@ -697,20 +697,23 @@ if(gnutls OR NOT OPENSSL_FOUND)
 	elseif(gnutls)
 		message(FATAL_ERROR "GnuTLS library not found")
 	endif()
-endif()
-
-if(NOT OPENSSL_FOUND)
-	find_public_dependency(LibGcrypt)
-	set_package_properties(LibGcrypt
-		PROPERTIES
-			URL "https://www.gnupg.org/software/libgcrypt/index.html"
-			DESCRIPTION "A general purpose cryptographic library"
-			TYPE RECOMMENDED
-			PURPOSE "Use GCrypt instead of the built-in functions for RC4 and SHA1"
-	)
-	if (LibGcrypt_FOUND)
-		target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_LIBGCRYPT)
-		target_link_libraries(torrent-rasterbar PRIVATE LibGcrypt::LibGcrypt)
+else()
+	if(TARGET OpenSSL::Crypto)
```

Comment:
You may want to use `libcrypto` (or `libgcrypt`, for that matter) even if `gnutls OR NOT TARGET OpenSSL::SSL`, right?

Reply:
I'm not sure `libcrypto` is independent of openssl actually. I think it openssl is enabled libcrypto may be implied. That's not the case with gnutls though

Reply:
>I think it openssl is enabled libcrypto may be implied. 

If OpenSSL is found it means, `libssl` (the `OpenSSL::SSL` target) and/or `libcrypto` (the `OpenSSL::Crypto` target) _may_ be available. If `libssl` is available, then `libcrypto` is also necessarily available, since `libssl` depends on it. But libcrypto can be available wihtout `libssl`.

Since it is my impression that there is code in libtorrent that can use `libcrypto` (or `libgcrypt`) functionality independently of the availability of `libssl` and `gnutls`, I think it makes sense to make those crypto features independent of the (non) availability of `libssl`/`gnutls`.

Reply:
right. I think this updated merge does that. But the longer term goal should be to make the configuration explicit, rather than depend on what libraries can be found

Reply:
To be clear: I think the outermost `if (NOT GNUTLS_FOUND AND NOT TARGET OpenSSL::SSL)` is not needed at all actually. This is because you may want to use `libcrypto`/`libgcrypt` (for `hasher.cpp`, for example), interdependently of `gnutls` or `libssl` being available.

Reply:
sure, but there is no way to express this wish in the CMakeList.txt right now, it's all prescribed based on what libraries are found. I think that should be fixed, but I don't think this merge changes that situation materially. Does it?

Also, I don't think libtorrent (currently) supports using OpenSSL *without* also using libcrypto. Maybe that should be changed, but it also doesn't seem reasonable to use some other crypto library since libcrypto is already linked in.

Reply:
@arvidn 

>I think that should be fixed, but I don't think this merge changes that situation materially. Does it?

yeah this is a change for another time.
 
> Also, I don't think libtorrent (currently) supports using OpenSSL _without_ also using libcrypto.

Yeah, but it does make sense to use `libcrypto` without `libssl`. Doesn't it?

>but it also doesn't seem reasonable to use some other crypto library since libcrypto is already linked in.

Even if you link against `gnutls`, wouldn't you maybe want to also link against `libcrypto` for the code in `hashser.cpp`, for example? Or can `gnutls` also handle that without `libcrypto` (or `libgcrypt`)?

Reply:
> Yeah, but it does make sense to use libcrypto without libssl. Doesn't it?

I agree.

> Or can gnutls also handle that without libcrypto (or libgcrypt)?

I used to be under the impression that gnutls was as tied to libgcrypt as openssl is to libcrypto, but that doesn't appear to be the case. So, yes, it might make sense to use gnutls and libcrypto, even though gnutls surely will also link in some other crypto library.